File: app/serializers/preview_serializer.rb
class PreviewSerializer < BaseSerializer
  attributes :title, :description, :url, :cover, :cover_srcset, :tags,
             :extra_context, :date_time

  alias_attribute :title, :preview_title
  alias_attribute :description, :formatted_description
  alias_attribute :cover, :cover_url

  delegate :url, :cover_srcset, :preview_title, :formatted_description,
           to: :object

  COLLECTION_TAG_LITERAL = 'Collection'.freeze

  def tags
    object.is_snippet? ? object.formatted_preview_tags : COLLECTION_TAG_LITERAL
  end

  def extra_context
    object.is_snippet? ? object.date_formatted : object.formatted_snippet_count
  end

  def date_time
    object.is_snippet? ? object.date_machine_formatted : nil
  end
end

File: src/serializers/previewSerializer.js
import Serializer from '#src/core/serializer.js';
import settings from '#src/config/settings.js';

export default class PreviewSerializer extends Serializer {
  static {
    Serializer.prepare(this, [
      ['title', 'previewTitle'],
      ['description', 'formattedDescription'],
      'url',
      ['cover', 'coverUrl'],
      'coverSrcset',
      [
        'tags',
        object => {
          return object.isSnippet
            ? object.formattedPreviewTags
            : settings.presentation.collectionPreviewTag;
        },
      ],
      [
        'extraContext',
        object => {
          return object.isSnippet
            ? object.dateFormatted
            : object.formattedSnippetCount;
        },
      ],
      [
        'dateTime',
        object => {
          return object.isSnippet ? object.dateMachineFormatted : null;
        },
      ],
    ]);
  }
}
